datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -798,24 +797,33 @@ private void issueAssignmentTokensForStoppingDatastreams(List<DatastreamGroup> s
         for (String instance : instances) {
           String assignmentTokenPath = KeyBuilder.datastreamAssignmentTokenForInstance(_cluster,
               stoppingStream.getName(), instance);
-          AssignmentToken token = new AssignmentToken(hostname, instance);
+          AssignmentToken token = new AssignmentToken(_instanceName, instance, System.currentTimeMillis());
           _zkclient.create(assignmentTokenPath, token.toJson(), CreateMode.PERSISTENT);
         }
       }
     }
   }
 
   /**
-   * Gets the name of the local host
+   * Claims assignment tokens for the given datastreams and instance
+   * @param datastreams List of datastreams for which tokens are to be claimed
+   * @param instance Instance name
    */
-  private String getLocalHostName() {
-    String hostname = "localhost";
-    try {
-      hostname = InetAddress.getLocalHost().getHostName();
-    } catch (UnknownHostException ex) {
-      LOG.warn("Unable to obtain hostname for leader");
+  public void claimAssignmentTokensForDatastreams(List<Datastream> datastreams, String instance) {
+    for (Datastream stream : datastreams) {
+      String streamName = stream.getName();
+      String tokenPath = KeyBuilder.datastreamAssignmentTokenForInstance(_cluster, streamName, instance);
+      if (_zkclient.exists(tokenPath)) {
+        if (instance.equals(_instanceName)) {
+          LOG.info("Claiming assignment token for datastream: {}", streamName);
+        } else {
+          LOG.info("Revoking assignment token for datastream: {}, instance: {}", streamName, instance);
+        }
+        _zkclient.delete(tokenPath);
```

Comment:
We should have a ZkException catcher to know if any updates fail.

Reply:
`_zkclient` doesn't throw an exception but it returns a boolean telling the caller if the delete was successful or not. It also logs a warning message if the delete fails:

```
public boolean delete(final String path) {
    long startT = System.currentTimeMillis();
    boolean var13 = false;

    boolean success;
    try {
      try {
        var13 = true;
        this.retryUntilConnected(new Callable<Object>() {
          public Object call() throws Exception {
            ZkClient.this.getConnection().delete(path);
            return null;
          }
        });
        success = true;
      } catch (ZkNoNodeException var14) {
        success = false;
        if (LOG.isDebugEnabled()) {
          LOG.debug("zkclient {}, Failed to delete path {}, znode does not exist!", this._uid, path);
        }
      }

      this.record(path, (byte[])null, startT, AccessType.WRITE);
      var13 = false;
    } catch (Exception var15) {
      this.recordFailure(path, AccessType.WRITE);
      LOG.warn("zkclient {}, Failed to delete path {}! ", new Object[]{this._uid, path, var15});
      throw var15;
    } finally {
      if (var13) {
        long endT = System.currentTimeMillis();
        if (LOG.isTraceEnabled()) {
          LOG.trace("zkclient {} delete, path: {}, time {} ms", new Object[]{this._uid, path, endT - startT});
        }

      }
    }
```
I don't think there's much we need to do here.

Reply:
The code you pasted does seem to throw the exception "throw var15;". And I was concerned here since there's no catch block in the code chain after that if that exception is thrown. 

And since the scheduled executor is running that thread, that executor thread would die, causing work stoppage, I guess if I am not wrong. Also, in this case, it does not harm as a new task to this executor would create a new thread, but catching that exception should make things simpler.

Reply:
Good catch. I didn't realize it's getting re-thrown. Added exception handling and error logging in ZkAdapter.